src/validators.ts

```diff
@@ -0,0 +1,135 @@
+import { Spec, ValidatorSpec } from './types'
+
+// Simplified adaptation of https://github.com/validatorjs/validator.js/blob/master/src/lib/isFQDN.js
+const isFQDN = (input: string) => {
+  if (!input.length) return false
+  const parts = input.split('.')
+  for (let part, i = 0; i < parts.length; i++) {
+    part = parts[i]
+    if (!/^[a-z\u00a1-\uffff0-9-]+$/i.test(part)) return false
+    if (/[\uff01-\uff5e]/.test(part)) return false // disallow full-width chars
+    if (part[0] === '-' || part[part.length - 1] === '-') return false
+  }
+  return true
+}
+
+// "best effort" regex-based IP address check
+// If you want a more exhaustive check, create your own custom validator, perhaps wrapping this
+// implementation (the source of the ipv4 regex below): https://github.com/validatorjs/validator.js/blob/master/src/lib/isIP.js
+const ipv4Regex = /^(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])$/
+const ipv6Regex = /([a-f0-9:]+:+)+[a-f0-9]+/
+const isIP = (input: string) => {
+  if (!input.length) return false
+  return ipv4Regex.test(input) || ipv6Regex.test(input)
+}
+
+const EMAIL_REGEX = /^[^@\s]+@[^@\s]+\.[^@\s]+$/ // intentionally non-exhaustive
+
+export class EnvError extends TypeError {
+  constructor(message?: string) {
+    super(message)
+    Error.captureStackTrace(this, EnvError)
+    this.name = 'EnvError'
+  }
+}
+
+export class EnvMissingError extends ReferenceError {
+  constructor(message?: string) {
+    super(message)
+    Error.captureStackTrace(this, EnvMissingError)
+    this.name = 'EnvMissingError'
+  }
+}
```

Comment:
I don't really remember why and where I found the solution but there was something with that I think I couldn't do `instanceof EnvMissingError` etc if I didn't do `Object.setPrototypeOf(this, EnvMissingError.prototype);`

reference from envsafe:
```ts
export class MissingEnvError extends ReferenceError {
  constructor(message?: string) {
    super(message);

    Object.setPrototypeOf(this, MissingEnvError.prototype);
  }
}
```

Reply:
Seems to work with the current implementation (via node repl):

<img width="352" alt="Screen Shot 2020-10-18 at 6 11 54 PM" src="https://user-images.githubusercontent.com/25241/96391426-78291d80-116d-11eb-9f3f-a1a5150c0458.png">

I'll add a test case to make sure the problem doesn't pop up in the future